sdk/cosmos/azure_data_cosmos/src/location_cache.rs

```diff
@@ -0,0 +1,570 @@
+            RequestOperation::Read,
+            &self.default_endpoint,
+        );
+
+        self.refresh_stale_endpoints();
+    }
+
+    fn get_endpoints_by_location(
+        &mut self,
+        locations: HashMap<String, String>,
+    ) -> (HashMap<String, String>, Vec<String>) {
+        // Separates locations into a hashmap and list
+        let mut endpoints_by_location: HashMap<String, String> = HashMap::new();
+        let mut parsed_locations: Vec<String> = Vec::new();
+
+        for (location, endpoint) in locations {
+            if !location.is_empty() {
+                endpoints_by_location.insert(location.clone(), endpoint.clone());
+                parsed_locations.push(location);
+            }
+        }
+
+        (endpoints_by_location, parsed_locations)
+    }
+
+    fn get_preferred_available_endpoints(
+        &self,
+        endpoints_by_location: &HashMap<String, String>,
+        request: RequestOperation,
+        default_endpoint: &str,
+    ) -> Vec<String> {
+        let mut endpoints: Vec<String> = Vec::new();
+        let mut unavailable_endpoints: Vec<String> = Vec::new();
+
+        for location in &self.locations_info.preferred_locations {
+            // Checks if preferred location exists in endpoints_by_location
+            if let Some(endpoint) = endpoints_by_location.get(location) {
+                // Check if endpoint is available, if not add to unavailable_endpoints
+                // If it is then add to endpoints
+                if !self.is_endpoint_unavailable(endpoint, request) {
+                    endpoints.push(endpoint.clone());
+                } else {
+                    unavailable_endpoints.push(endpoint.clone());
+                }
+            }
+        }
+
+        // Add unavailable endpoints to end of endpoints lists
+        for endpoint in unavailable_endpoints {
+            endpoints.push(endpoint);
+        }
+
+        // If no preferred locations were found, use the default endpoint
+        if endpoints.is_empty() {
+            endpoints.push(default_endpoint.to_string());
+        }
+
+        endpoints
+    }
+
+    fn refresh_stale_endpoints(&mut self) {
+        let mut location_unavailability_info_map =
+            self.location_unavailability_info_map.lock().unwrap();
+
+        // Removes endpoints that have not been checked in the last 5 minutes
+        location_unavailability_info_map.retain(|_, info| {
+            info.last_check_time.elapsed().unwrap_or_default()
+                <= std::time::Duration::from_secs(300)
+        });
+    }
+}
+
+// Tests for location cache
+#[cfg(test)]
+
+mod tests {
+    use super::*;
+    use std::{collections::HashSet, vec};
+
+    fn create_test_data() -> (
+        String,
+        HashMap<String, String>,
+        HashMap<String, String>,
+        Vec<String>,
+    ) {
+        // Setting up test database account data
+        let default_endpoint = "https://default.documents.azure.com".to_string();
```

Comment:
nit: I like to use `.example.com` as the domain name whenever I'm writing test URLs that we don't actually send requests to. The entire `example.com` domain is reserved exactly for this purpose and it means we're never in danger of accidentally referencing a real endpoint. So, for this, I'd use `https://default.documents.example.com`.